Service Catalog and Broker Removal

jmrodri opened the pull request.

deads2k commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,187 @@
+Ansible Service Broker, Template Service Broker, or any broker registered to
+Service Catalog.
+
+During the upgrade, alert users if the Service Catalog or any Service Brokers
```
During the upgrade seems too late to make a decision.  Back-pressure would be great because it would allow a cluster-admin to make a controlled choice.  We have a similar choice to be made for handling removal of resources from kube.

jmrodri replied:
There was mention of using a Job in the manifests directory to accomplish the removal but I'm not 100% sure of the details yet.

https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/

jmrodri replied:
@deads2k what do you mean by "back-pressure"?

derekwaynecarr replied:
can we add warnings in 4.3.z if the capability is in use leading up to 4.4 release?

jmrodri replied:
@derekwaynecarr what sort of warnings? In 4.3, if the service catalog or brokers are enabled, a prometheusrule is created to fire an alert that they are enabled.

deads2k commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,187 @@
+
+#### Story 1
+
+Create a pre-flight check in the CVO process to ensure the Service Catalog,
```
I'm not aware of existing "pre-flight" checks that are generically injected by clusteroperators.  Do you have a link for how this works today or is the idea net new?

jmrodri replied:
There was mention of using a Job in the manifests directory to accomplish the removal but I'm not 100% sure of the details yet.

https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/

deads2k commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,187 @@
+   Service Brokers are not present after an upgrade to OpenShift 4.4. (ENG)
+1. Alerts users during upgrade if Service Catalog and Service Brokers are in
+   use. (ENG)
+1. Prevent users from upgrading to OpenShift 4.4 if the Service Catalog or the
```
this isn't explored below, but seems valid.

deads2k commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,187 @@
+
+#### Story 2
+
+Remove Service Catalog and the brokers from the release payload
```
Seems like one simple option to handling this is to leave the service catalog operator present in 4.4, but update it to always remove its resources.  This would ensure everything is cleaned up and the CVO is only responsible for cleaning resources the CVO created.

jmrodri replied:
Updated proposal to include keeping the operators present.

deads2k commented:
If you had real backpressure, you could push all cleanup to happen prior to upgrading to 4.4, which would alleviate most of your upgrade and decommisioning concerns. It would make the responsibility of handling orphaning versus deprovision a straight cluster-admin decision.

derekwaynecarr commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,187 @@
+
+## Motivation
+
+Service Catalog and Service Brokers need to be removed in OpenShift 4.4 because
```
4.4 is not necessarily the LTS, the motivation is to just finish the previously announced deprecation.

jmrodri replied:
Removed reference to LTS.

deads2k commented:
> If you had real backpressure, you could push all cleanup to happen prior to upgrading to 4.4, which would alleviate most of your upgrade and decommisioning concerns. It would make the responsibility of handling orphaning versus deprovision a straight cluster-admin decision.

A more complete alternative pitch.
 1. update 4.3.z to have the service catalog set upgradeable=false if managementstate is managed. 
 You can have a message explaining why and linking to docs.
 2. The directions need to explain their deletion/orphaning options to them and ideally provide them some example oc commands to help them write their own scripts.  Then advise the user to either manage service catalog themselves with no help (unmanaged) or remove the api entirely which results in removing everything created by it and putting it into removed.

If the state is unmanaged, then no further action needs to be taken because the cluster-admin takes responsibility for managing service catalog on an ongoing basis.  This means that any service catalog resources remaining are the cluster-admin's responsibility.

If the state is removed, then no further action needs to be taken because there are no more service catalog resources present in the cluster.

The only change I see needed is a way to distinguish "may not upgrade to 4.y+1" versus "may not upgrade to 4.y.z+1" and that update to the CVO has value for all operators on an ongoing basis.

jmrodri commented:
@deads2k @derekwaynecarr I updated the proposal to be a bit more clear. There is definitely still some ambiguity in exactly how we solve this.

jmrodri commented:
> The only change I see needed is a way to distinguish "may not upgrade to 4.y+1" versus "may not upgrade to 4.y.z+1" and that update to the CVO has value for all operators on an ongoing basis.

We were going to make `Upgradeable=False` in 4.3 but this exact scenario you describe here caused us to keep it at True. We didn't want to block an upgrade to 4.3.z but we did want to block upgrading to 4.4.

joelanford commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,211 @@
+
+We could let the upgrade occur and change the behavior of the existing operators
+to simply remove the objects they own. If we go this route, we will need to keep
+the operators in the release payload.
```
IIUC, it seems like this approach would need the associated CRDs to be present in the cluster and the operators running. Would these things self-destruct once the upgrade is complete, or would they remain in the cluster?  If the latter, that might be confusing to users and other processes that use the kubernetes discovery API.

Would we need a validating admission webhook to block the creation of new CRs?

jmrodri replied:
Updated the proposal to use a Job https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/

This should mean the CRDs can go away.

joelanford commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,211 @@
+### Risks and Mitigations
+
+The biggest risk is if there needs to be additional work done to the CVO in
+order to support anything we would like to do described in this proposal.
```
Any chance we can close the loop on the CVO questions this week to better understand these risks?

jmrodri replied:
Removed, CVO is not likely to have any major changes.

joelanford commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,211 @@
+#### Story 1
+
+Implement a mechanism that ensures the Service Catalog and Service Brokers are
+not present after an upgrade to OpenShift 4.4.
```
Option 1 (block the upgrade) and option 2 (automatically remove) have some pretty significant differences from a UX standpoint. Is there a preference about which approach we should take? 

It seems like option 1 would be vastly preferable to force the decision on the cluster admin.

Is option 2 something we would seriously consider?

wking commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,211 @@
+We could potentially block upgrades from happening until the admin removes the
+Service Catalog and the brokers. This would need a mechanism that blocks
+upgrades. There is an `Upgradeable: False` in the operator's status but not sure
+if that is checked by the CVO during upgrades.
```
CVO should be checking this since openshift/cluster-version-operator#243

jmrodri replied:
@wking if I set `Upgradeable` to false, will that prevent major upgrades, i.e. 4.3 -> 4.4? What about z-stream, i.e. 4.3.0 -> 4.3.1? Presumably I'd like to allow upgrades in the z-stream, but not going to 4.4.

wking commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,211 @@
+
+1. Remove the Service Catalog from productization (ART)
+1. Remove the installation instructions from the documentation (Docs)
+1. Ensure you how to disable service catalog instructions are correct (Docs/Eng)
```
nit: "you" -> "the"

jmrodri replied:
Fixed.

shawn-hurley commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,210 @@
+1. Remove the Service Catalog from the OpenShift 4.4 release payload (ART)
+   * Will have to keep the Service Catalog operators if we choose to repurpose
+     them.
+1. Remove the Ansible Service Broker from the OpenShift 4.4 release payload (ART)
```
These are not in the release payload

jmrodri commented:
@wking @shawn-hurley @joelanford @deads2k @derekwaynecarr I reworked the entire proposal to use a  [Job](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/) to handle removal of the Service Catalog. I also added several alternatives in the Alternatives section.

shawn-hurley commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,266 @@
+This might be accomplished with just one of the operators, that would need
+a little more investigation.
+
+Replace the operator's code with a new Job executable that will perform the
```
We need to make sure that this orphan's everything. I thought we talked about removing these resources be removing the API server and *not* actually deleting these resources, correct?

jmrodri replied:
That was done in either 4.2 or 4.3, but we remove the owner references and orphan things. So we're good there.

shawn-hurley commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,266 @@
+
+1. Keep the Service Catalog _Operators_ in the OpenShift 4.4 release payload
+1. Remove the Service Catalog from the OpenShift 4.4 release payload (ART)
+1. Remove the Ansible Service Broker from the OpenShift 4.4 release payload (ART)
```
These are not apart of the payload

jmrodri replied:
Addressed.

shawn-hurley commented on enhancements/service-catalog/service-catalog-and-broker-removal.md:
```diff
@@ -0,0 +1,266 @@
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+* what about Custom Broker installed, how will those be handled?
```
They stay running they just don't do anything. I think the same should be said for ASB and TSB IMO we should not remove them just orphan them

jmrodri replied:
added answer to the doc

joelanford commented:
/lgtm
